### PR TITLE
data-tiling: introduce `upper_bound_tile_size` op to defer padding-size choice to MaterializeEncoding.

### LINKED ISSUE #11632: Avoid unnecessary padding in `iree-flow-set-encoding`
(Updated July 2023)

When data-tiling, during Flow we don't know yet the exact tile sizes, so we don't know yet exactly how much padding is needed to ensure that buffers are allocated large enough to account for padding to the next multiple of what will be the actual tile size.

The current code is simply using a hardcoded tile size of 16, along every dimension, in Flow.

Problems with that:
1. That's assuming that no tile size will ever be larger than 16, which is not a safe assumption with ever-wider SIMD. Intel AMX, Apple AMX, ARM SME, SiFive (RISC-V) Intelligence extensions are all defeating that. In fact, with narrow data types such as `float16`, even AVX-512 or ARM SVE or RISC-V RVV can already defeat that with tile sizes equal to 32. So _resolving that blocks being able to target all these near-team SIMD targets_.
2. When the matrix shapes are very narrow, meaning that some dimension size is less than 16, padding to the next multiple of 16 can be large overhead. Particularly for matrix-times-vector multiplications, the widening is from 1 to 16 columns, immediately rephrasing the problem at hand into one that involves 16x more arithmetic.
3. Incomplete tiles are costly to deal with in packing code. We would normally have at most one incomplete tile at the end of each dimension, but because of this, we can have as many as 16.
4. This requires the `SetEncoding` pass to be parametrized in a `defaultPadding` option, while it could otherwise be unparametrized.

### PR DESCRIPTION
**Pull request by bjacob**
This fixes #11632, by introducing a materializable `upper_bound_tile_size ` instead of hardcoding a fixed padding amount at Flow, and fixes it in sufficient generality to also solve the problem for narrow matmuls - let's explain that in more detail as this is an important part of what this PR is doing.

For each combination of element types and each target, the MaterializeEncoding pass selects appropriate matmul tile shapes. Input tensors get padded to the next multiple of the next tile size. The padding increases the inherent arithmetic cost of the problem at hand. When, along some dimension, the original tensor size is smaller than the tile size, that can result in particulary large overhead. The extreme case, which is also a very common case, is matrix-times-vector multiplication. The "vector" right-hand side is really a matrix with one dimension size equal to 1, so if the general matmul tile shape along that dimension is 8 or 16, as is usually the case, that can be a 8x or 16x increase in the inherent arithmetic cost of the matmul op.

The solution to that is to adjust MaterializeEncoding tile shapes to narrow dimensions. We had some logic in place to deal with that, but #11632 was leaving it moot: the flow-level padding of everything to the next multiple of 16 meant that our logic there never really had a chance of kicking in.  With #11632 being fixed, this PR was the opportunity to also fix that along the way, and to ensure that the solution to #11632 worked also in that respect. As matrix-times-vector products were the common case that suffered the most from #11632, it would have been too bad to "solve" #11632 without addressing that. By the way, matrix-times-vector is only the extreme case, but other narrow cases matter too. When, e.g. on AVX-512, the general matmul tile size is 16, even width-8 matmuls (MxKx8) were suffering from 2x-widening. So the solution in this PR is making sure to address all narrow cases, defined as whenever a tensor dimension size is less than the general tile size.

The difficulty was that when MaterializeEncoding runs on a dispatch function, it runs on an already-padded tensor; even as this PR introduces `upper_bound_tile_size`, that only makes it possible to select the right padding amount, but there's still a `tensor.pad` op and it's still getting in the way of knowing the actual, original tensor shape for the purpose of adjusting tile shapes for narrow cases. Moreover, as `MaterializeEncoding` is a type-converter pass, it can't just walk from a Value up to its defining-op to find the pre-padding tensor. There are no values there, only types. So the information about the pre-padding tensor shape has to be part of the tensor type that `MaterializeEncoding` sees, that its, the padded tensor type.

The solution to that problem in this PR is to add a `original_type` field to `EncodingAttr`.

Fixes  #11632.

Fixes a compiler issue encountered in #14398 but not the originally reported runtime crash by itself.

This now also includes the removal of a now-useless VMVX pass, which was originally split into https://github.com/openxla/iree/pull/14383 .

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/test/set_encoding.mlir
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/test/set_encoding.mlir]: can you expand this to `original_type`? (abbreviations aren't generally very friendly to those without familiarity with the context or language)

File: llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td
Comment (benvanik) [inline llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td]: having `tensor<> <>` reads weird - you can make this `qualified($encoding)` so it prints out the attr name and makes more sense. it's useful to add some joining word/preposition between to indicate to readers what the attribute represents, as well, such as `in` or `as` ("the **tensor of this size/element type** is **encoded as blah** returning **a set of dimension values**")

File: llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtBase.td
Comment (benvanik) [inline llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtBase.td]: `TypeAttrOf<AnyRankedTensor>` can be used instead of defining a new type (fewer defs the better)

File: llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td
Comment (benvanik) [inline llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td]: TODO

**Review by benvanik — CHANGES_REQUESTED** [review]

File: llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td
Comment (bjacob) [reply llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td]: thanks, i didn't know about `qualified`.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/SetEncoding.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/SetEncoding.cpp:43]: This is the part of the PR that I think was being discussed. Following Ben's suggestion about teaching stream to use the logical shape + tensor encoding [here](https://github.com/openxla/iree/blob/3f71a2f7f95971cc28d2219428b751050343d62f/compiler/src/iree/compiler/Dialect/Stream/Transforms/EncodeTensors.cpp#L49) will make it unnecessary to use an explicit `pad` operation. We will basically need only to add a `iree_linalg_ext.set_encoding` cause the encoding effectively includes padding at that point. So then that will obviate the need for the op added here. I am sorry I suggested that we should add an op but that was under the assumption that we had that stream needs the logical shape and physical shape to correspond exactly. But Ben is suggesting that we can use logical shape + encoding to get the physical shape. That is the more robust way forward. I see there is a lot of code here that you added to make this work, but all of that might become unnecessary with Ben's suggestion. That's my mistake (I was working off prior from discussing with Ben a few months ago)

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Still working through the review. This is a bit dump that is touching a lot of things... left a comment based on discussion on Discord and with Ben.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUMaterializeEncodingPass.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUMaterializeEncodingPass.cpp]: I think you made this a separate pass because you want to iterate over a array of targets and pick a pad size? I'd assume that you would do the same for the actual inner tile dimensions of the packing operation. So this seems a bit premature. Maybe we can drop this extra pass and just incorporate this into the existing pass?

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/SetEncoding.cpp
Comment (bjacob) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/SetEncoding.cpp:43]: I think we agree 100% - but my understanding of the outcome of yesterday's discussion is that even though this isn't the final state, it is useful to merge this, as:

- While this adds some code that will be thrown away, it replaces an equivalent mass of preexisting code (particularly with the follow-up PR) and it brings the design closer to what Ben wants to replace it with. It replaces some brittle hacks (the SetEncoding pass taking a defaultPadding option with a default value 16 that must be left in its default value or else mysterious crashes happen at runtime as padded buffers overflow).
- This unblocks multiple conversations (narrow matrix multiplication performance, writing microkernels for these cases).

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUMaterializeEncodingPass.cpp
Comment (bjacob) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUMaterializeEncodingPass.cpp]: The primary reason why this needs to be a separate pass is that MaterializeEncoding runs on HAL executables (dispatch functions, which are where `set_encoding` ops are found at that point), while this needs to run on "top-level" code that is outside of HAL executables (the function that makes the flow.dispatch calls).

While these `upper_bound_tile_size` ops are initially found on both sides (both inside and outside of HAL executables), the ones that are in HAL executables are just dropped as the `pad + set_encoding` pairs of ops get rewritten into just `tensor.pack`, the `pad` op goes unused, and it was the only thing there that was using the `upper_bound_tile_size`).  By contrast, the `upper_bound_tile_size` on the outside of HAL executables are consumed by flow.dispatch ops, so they don't go away, so that's where we need this pass to run on them.

Then the aspect that this needs to iterate over an array of targets, is just a consequence of that -- it's only inside of a HAL executable that one knows one specific target.

It's still possible that one could find a way to unify these two passes, just saying why it's not completely trivial, it would bring more refactoring into the scope of this PR.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/SetEncoding.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/SetEncoding.cpp:63]: Nice! +1 for using `composedFoldedAffineApply`.

File: compiler/src/iree/compiler/Codegen/Common/DecomposePackUnPackOps.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/DecomposePackUnPackOps.cpp:89]: I dont think I follow this one. Can you just provide some context here (just for my information).

**Comment by MaheshRavishankar** [comment]
Ok, after some offline conversation, the conclusion was that this gets us to a better state from where we were, and a follow up will help make some of this unnecessary. There are a lot of clean ups here that we might as well bank in the intermediate.

@benvanik This looks OK to me. Do you have any thoughts?

**Review by MaheshRavishankar — APPROVED** [review]

File: compiler/src/iree/compiler/Dialect/HAL/Transforms/Passes.cpp
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/HAL/Transforms/Passes.cpp]: this should be
`FunctionLikeNest(passManager).addPass(createYourPass);`
other ops (like util.initializer) may contain dispatch logic

File: compiler/src/iree/compiler/Dialect/HAL/Transforms/Passes.cpp
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/HAL/Transforms/Passes.cpp]: this really bothers me (HAL should not be pulling in target-specific passes) - is there a reason this has to be LLVMCPU, or can it live in common? it's confusing that it is used by VMVX (which is not LLVMCPU) and would be worth renaming

**Review by benvanik — CHANGES_REQUESTED** [review]
Yeah I think this is fine to go in with the cleanups fast-following - really don't want this kind of pass to exist outside of codegen pipelines but there's some really good stuff here and I think we all have visibility into the next steps. Once this lands we should regroup on an issue and make sure we're tracking what's next. I've been itching to have use of the runtime device queries/executable constant blocks/etc and this will be the perfect thing to pull all that together.

before landing I would like to not have LLVMCPU be pulled in to the HAL pipeline - that's a big layering violation - moving the pass to Common/ would sketch me out much less (even though in the follow on we should remove the pass entirely).

**Comment by bjacob** [comment]
Incidentally this happens to also fix #14398.

**Comment by stellaraccident** [comment]
> Incidentally this happens to also fix #14398.

Thank you for this. Just happens to fix or do we know why? (not looking a gift horse in the mouth, but been chasing some things that may have been related)

**Comment by stellaraccident** [comment]
> > Incidentally this happens to also fix #14398.
> 
> Thank you for this. Just happens to fix or do we know why? (not looking a gift horse in the mouth, but been chasing some things that may have been related)

nm - I see you commented on that bug. Thank you for being on top of it.

File: compiler/src/iree/compiler/Codegen/Common/DecomposePackUnPackOps.cpp
Comment (bjacob) [reply compiler/src/iree/compiler/Codegen/Common/DecomposePackUnPackOps.cpp:89]: Sure, see this gist with a detailed testcase + error + ir-after-all log when this change is omitted.
https://gist.github.com/bjacob/42ea8ebbfafa780f030c0f5450ce4209

File: llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td
Comment (ScottTodd) [inline llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td:1075]: I just landed some changes to the structure of this LinalgExtOps.td file in https://github.com/openxla/iree/pull/14388 . Before merging, could you sync and ensure that this op is in the right `OpDocGroup` / part of the file? (OpGroupDataTilingOps?)

```c++
//===----------------------------------------------------------------------===//
// Data tiling ops
//===----------------------------------------------------------------------===//

def OpGroupDataTilingOps : OpDocGroup {}

let opDocGroup = OpGroupDataTilingOps in {

// YOUR OP SHOULD BE HERE? //

} // OpGroupDataTilingOps
```

File: compiler/src/iree/compiler/Dialect/HAL/Transforms/Passes.cpp
Comment (bjacob) [reply compiler/src/iree/compiler/Dialect/HAL/Transforms/Passes.cpp]: I tried it, but got this runtime error:

```
LLVM ERROR: Can't add pass 'LLVMCPUMaterializeEncoding' restricted to 'func.func' on a PassManager intended to run on 'util.initializer', did you intend to nest?
```

File: llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td
Comment (bjacob) [reply llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.td:1075]: Yup, checked, it does belong in there. Thanks!

File: compiler/src/iree/compiler/Dialect/HAL/Transforms/Passes.cpp
Comment (bjacob) [reply compiler/src/iree/compiler/Dialect/HAL/Transforms/Passes.cpp]: Moved to Codegen/Common/CPU.

**Comment by bjacob** [comment]
Updated to break the dependency of HAL on LLVMCPU by creating Codegen/Common/CPU.  And folded into this the removal of VMVXMaterializeEncodingPass, which was formerly split into https://github.com/openxla/iree/pull/14383.

File: compiler/src/iree/compiler/Dialect/HAL/Transforms/Passes.cpp
Comment (MaheshRavishankar) [reply compiler/src/iree/compiler/Dialect/HAL/Transforms/Passes.cpp]: I think the issue is that the pass is defined as a pass over `func.func` instead of `FuncOpInterface`. You need to define the pass similar to this https://github.com/openxla/iree/blob/093afb78ed3258cfc6d072da37a5ec184bb89f22/compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.td#L65. The it can apply to any op that implements the interface.

File: compiler/src/iree/compiler/Codegen/Common/CPU/Passes.td
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/CPU/Passes.td:16]: Ultra nit: Could you add some comments similar to https://github.com/openxla/iree/blob/093afb78ed3258cfc6d072da37a5ec184bb89f22/compiler/src/iree/compiler/Codegen/Common/GPU/Passes.td#L12

**Review by MaheshRavishankar — APPROVED** [review]

File: compiler/src/iree/compiler/Dialect/HAL/Transforms/Passes.cpp
Comment (bjacob) [reply compiler/src/iree/compiler/Dialect/HAL/Transforms/Passes.cpp]: That works, thanks!

File: compiler/src/iree/compiler/Codegen/Common/CPU/Passes.td
Comment (bjacob) [reply compiler/src/iree/compiler/Codegen/Common/CPU/Passes.td:16]: It's not a nit. Those `=` and `-` characters are converted into heat by the C preprocessor, which can be harnessed to produce energy. The `=` character actually produces 2x more heat than the `-`. At least that was the case in the original implementation of ASCII, which was made of wood.

**Review by benvanik — APPROVED** [review]

**Comment by bjacob** [comment]
Wait, disabling auto-merge: this PR actually makes the build_e2e_test_artifacts CI job take 3 hours, vs 10 minutes normally.  Need to investigate that before merging. @MaheshRavishankar @hanhanW @benvanik

**Comment by pzread** [comment]
> Wait, disabling auto-merge: this PR actually makes the build_e2e_test_artifacts CI job take 3 hours, vs 10 minutes normally. Need to investigate that before merging. @MaheshRavishankar @hanhanW @benvanik

Tried to reproduce locally and found this one takes a long time
```
iree-compile --output-format=vm-bytecode \
  --mlir-print-op-on-diagnostic=false \
  --iree-hal-target-backends=llvm-cpu --iree-input-type=tosa \
  --iree-llvmcpu-target-triple=aarch64-none-linux-android29 \
  --iree-flow-enable-data-tiling \
  --iree-llvmcpu-target-cpu-features=+dotprod \
  --iree-flow-enable-fuse-padding-into-linalg-consumer-ops \
  --iree-llvmcpu-enable-pad-consumer-fusion \
  iree_MobileBertSquad_int8_tflite_.mlir \
  -o iree_module_MobileBertSquad_int8_tflite___armv8.2-a-generic-linux_android29-llvm_cpu__experimental-flags_mmt4d_dotprod_/module.vmfb \
  --iree-hal-executable-object-search-path="/usr/local/google/home/cheyuw/iree/iree2-build" \
  --iree-llvmcpu-embedded-linker-path="/usr/local/google/home/cheyuw/iree/iree2-build/llvm-project/bin/lld" \
  --iree-llvmcpu-wasm-linker-path="/usr/local/google/home/cheyuw/iree/iree2-build/llvm-project/bin/lld"
```

Imported MLIR can be grab from https://storage.googleapis.com/iree-github-actions-presubmit-artifacts/5555908660/1/e2e-test-artifacts/iree_MobileBertSquad_int8_tflite_.mlir

Or see https://github.com/openxla/iree/blob/main/docs/developers/developing_iree/benchmark_suites.md#build-benchmark-suites to import model and build benchmark suite locally

**Comment by bjacob** [comment]
Pushed a fix for the accidental materialization of dynamic shapes: https://github.com/openxla/iree/pull/14349/commits/e56d1c3b44e04453ed5709414bd2cbe0fe792366

adding the missing static-shape testcase to materialize_encoding.mlir.

**Comment by bjacob** [comment]
> Tried to reproduce locally and found this one takes a long time

Thanks @pzread ... actually, the one you identified here is still taking a long time even after the latest fix, so this still blocks merging this PR.

**Comment by bjacob** [comment]
Debugged the long compile time... testing a fix at https://github.com/openxla/iree/pull/14412 ... will post the gory details there. Rebased this PR on top of that so we should see CI e2e_artifacts completing fast again now here.

**Comment by bjacob** [comment]
Closing the loop on https://github.com/openxla/iree/pull/14349#issuecomment-1636299135 - the compile time regression - and the fix that I earlier tested as #14412 but decided to fold into this PR instead as doing #14412 on its own runs into issues that are precisely what this PR fixes (see closing comment there).

Minimizing the failing test from https://github.com/openxla/iree/pull/14349#issuecomment-1636299135 yields:

```mlir
func.func @main(%1092: tensor<384x32xi8>, %1091: tensor<384x32xi8>)
    -> (tensor<384x384xi32>) {
  %83 = "tosa.const"() <{value = dense<0> : tensor<384xi32>}> : () -> tensor<384xi32>
  %1093 = "tosa.fully_connected"(%1092, %1091, %83) <{quantization_info = #tosa.conv_quant<input_zp = 4, weight_zp = 3>}> : (tensor<384x32xi8>, tensor<384x32xi8>, tensor<384xi32>) -> tensor<384x384xi32>
  return %1093 : tensor<384x384xi32>
}
```

What's so special about this quantized matmul from this specific model, is the `weight_zp = 3`. We initially designed TFLite's [original quantization scheme](https://arxiv.org/abs/1712.05877) with arbitrary zero-points for both input and weights; then this was superseded by a [newer quantization scheme](https://www.tensorflow.org/lite/performance/quantization_spec) where the weights zero-point is zero ("symmetric quantization of weights"). So that's why few current models exercise this.

The `input_zp` plays no role in triggering the issue here because the lowering to `linalg.matmul` + generics only involves generics on the constant weights. By contrast, the nonzero `weights_zp` [causes `linalg.generic` ops](https://github.com/openxla/iree/blob/49335b7ee56d3d3b58b0850a3dbba3e179524739/compiler/src/iree/compiler/InputConversion/Common/QuantizedMatmulToMatmul.cpp#L91-L96) to compute a partial sum reduction of the input tensor, multiply that by `weights_zp`, and add that to the output of the `tensor.unpack` that is produced by materializing the data-tiled `tosa.fully_connected`.

The new commit in this PR, https://github.com/openxla/iree/pull/14349/commits/17a975482dd35ba0b6a796a5d18e3cf95261e824 , restores a fusion there, that IIUC was always intended but was brittle --- we were fusing `extract_slice` into `unset_encoding`, but in a way that still allowed `extract_slice` to act as a barrier preventing further fusions.

**Comment by bjacob** [comment]
@MaheshRavishankar since you already approved, and the newly added commit in this PR, https://github.com/openxla/iree/commit/17a975482dd35ba0b6a796a5d18e3cf95261e824, is tiny and seems to make sense to me (see https://github.com/openxla/iree/pull/14349#issuecomment-1637306000), and the above benchmark results are looking good, with in particular the dispatch count reductions confirming that it's seeing the effect of this commit... I'm going to merge this now, but we can always revisit if there's something objectionable in that last commit.

**Pull request merged**
